cdc/kv/regionlock/region_range_lock.go

```diff
@@ -31,84 +31,194 @@ import (
-	m *btree.BTreeG[*rangeTsEntry]
+	m     *btree.BTreeG[rangeTsEntry]
+	start []byte
+	end   []byte
 }
 
 // newRangeTsMap creates a RangeTsMap.
 func newRangeTsMap(startKey, endKey []byte, startTs uint64) *rangeTsMap {
 	m := &rangeTsMap{
-		m: btree.NewG(16, rangeTsEntryLess),
+		m:     btree.NewG(16, rangeTsEntryLess),
+		start: startKey,
+		end:   endKey,
 	}
-	m.Set(startKey, endKey, startTs)
+	m.set(startKey, endKey, startTs)
 	return m
 }
 
-// Set sets the corresponding ts of the given range to the specified value.
-func (m *rangeTsMap) Set(startKey, endKey []byte, ts uint64) {
-	if _, ok := m.m.Get(rangeTsEntryWithKey(endKey)); !ok {
-		// To calculate the minimal ts, the default value is math.MaxUint64
-		tailTs := uint64(math.MaxUint64)
-		m.m.DescendLessOrEqual(rangeTsEntryWithKey(endKey), func(i *rangeTsEntry) bool {
-			tailTs = i.ts
+func (m *rangeTsMap) clone() (res *rangeTsMap) {
+	res = &rangeTsMap{
+		m:     btree.NewG(16, rangeTsEntryLess),
+		start: m.start,
+		end:   m.end,
+	}
+	m.m.Ascend(func(i rangeTsEntry) bool {
+		res.m.ReplaceOrInsert(i)
+		return true
+	})
+	return
+}
+
+func (m *rangeTsMap) set(startKey, endKey []byte, ts uint64) {
+	if !bytes.Equal(m.end, endKey) && !m.m.Has(rangeTsEntryWithKey(endKey)) {
+		var found bool
+		var endKeyOverlapped rangeTsEntry
+		m.m.DescendLessOrEqual(rangeTsEntryWithKey(endKey), func(i rangeTsEntry) bool {
+			found = true
+			endKeyOverlapped.ts = i.ts
+			endKeyOverlapped.isUnset = i.isUnset
 			return false
 		})
-		m.m.ReplaceOrInsert(&rangeTsEntry{
-			startKey: endKey,
-			ts:       tailTs,
-		})
+		if found {
+			if !endKeyOverlapped.isUnset {
+				log.Panic("rangeTsMap double set")
+			}
+			endKeyOverlapped.startKey = endKey
+			m.m.ReplaceOrInsert(endKeyOverlapped)
+		}
 	}
 
-	entriesToDelete := make([]*rangeTsEntry, 0)
-	m.m.AscendRange(rangeTsEntryWithKey(startKey), rangeTsEntryWithKey(endKey),
-		func(i *rangeTsEntry) bool {
-			entriesToDelete = append(entriesToDelete, i)
-			return true
+	if !m.m.Has(rangeTsEntryWithKey(startKey)) {
+		var found bool
+		var startKeyOverlapped rangeTsEntry
+		m.m.DescendLessOrEqual(rangeTsEntryWithKey(startKey), func(i rangeTsEntry) bool {
+			found = true
+			startKeyOverlapped.isUnset = i.isUnset
+			return false
 		})
+		if found && !startKeyOverlapped.isUnset {
+			log.Panic("rangeTsMap double set")
+		}
+	}
 
+	entriesToDelete := make([]rangeTsEntry, 0)
+	m.m.AscendRange(rangeTsEntryWithKey(startKey), rangeTsEntryWithKey(endKey), func(i rangeTsEntry) bool {
+		if !i.isUnset {
+			log.Panic("rangeTsMap double set")
+		}
+		entriesToDelete = append(entriesToDelete, i)
+		return true
+	})
 	for _, e := range entriesToDelete {
 		m.m.Delete(e)
 	}
 
-	m.m.ReplaceOrInsert(&rangeTsEntry{
-		startKey: startKey,
-		ts:       ts,
+	m.m.ReplaceOrInsert(rangeTsEntry{startKey: startKey, ts: ts})
+}
+
+func (m *rangeTsMap) unset(startKey, endKey []byte) {
+	var neighbor rangeTsEntry
+	var exist bool
+
+	if !bytes.Equal(m.end, endKey) {
```

Comment:
IMOP, it is better to check `if bytes.Equal(m.end, endKey)`  first to reduce the nested if blocks.